### ISSUE TITLE
Like button in adapter usage

### ISSUE THREAD

**Issue by salihyalcin** [jd-alexander/LikeButton#2]
Hello :) dear developer

Your library makes my project's favourite functionality so powerful than mine's basic functionality

But I couldn't afford to add animations when using liked. It only works with changing background color property. For example; When I click the star button, only start button's background color became yellow :S Another animation not showing when I click the start button :S:S (I'd defined both xml and java side)

Here is my approach

XML;
```
  <com.like.LikeButton
        app:icon_type="star"
        app:icon_size="25dp"
        android:id="@+id/favorites"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content"
        android:layout_alignParentEnd="true"
        android:layout_alignParentRight="true"
        android:layout_marginRight="10dp"
        android:layout_marginEnd="10dp"
        android:layout_marginTop="10dp"
        app:circle_start_color="@color/colorPrimary"
        app:circle_end_color="@color/colorAccent"
        app:dots_primary_color="@color/colorAccent"
        app:dots_secondary_color="@color/colorPrimary"
        app:liked="false"/>
```

Java

```
if (!mStates.get(pos)) {
               
                    holder.mFavourite.setLiked(true);         
                    holder.mFavourite.setCircleEndColorRes(R.color.colorAccent);
                    holder.mFavourite.setExplodingDotColorsRes(R.color.colorPrimary,R.color.colorAccent);
                    mStates.put(pos, true);
                } else {           
                    holder.mFavourite.setLiked(false);
                    mStates.put(pos, false);
                }
```
TL;DR -  I could not afford to run like button with animation

**Comment by jd-alexander** [comment]
Okay give me some more information. Just state specifically what the issue is.  I'm trying to understand what you are saying but not sure. Would be good if you could upload an example to anorther repo so  I could try and reproduce the error; hoping that won't be necessary though.

**Comment by jd-alexander** [comment]
You can only trigger the animation from code. It can't be triggered by using the setLiked(true) function. That only changes the colour of the icon to a liked state.

**Comment by jd-alexander** [comment]
When you say you clicked the star button do you mean? actually tapping it on the screen because using the set liked function as I noted does not perform the animation.

**Comment by salihyalcin** [comment]
Thank you for fast turn back. 

 I made video :) 
https://www.youtube.com/watch?v=1BiThOLlKuc&feature=youtu.be

I'm letting video to speak instead of me :))

**Comment by salihyalcin** [comment]
As I said before I use it in java ?

```
holder.mFavourite.setCircleEndColorRes(R.color.colorAccent);
holder.mFavourite.setExplodingDotColorsRes(R.color.colorPrimary,R.color.colorAccent);
```

**Comment by jd-alexander** [comment]
Okay I saw the video. When is this code being triggered? 

if (!mStates.get(pos)) {

                    holder.mFavourite.setLiked(true);         
                    holder.mFavourite.setCircleEndColorRes(R.color.colorAccent);
                    holder.mFavourite.setExplodingDotColorsRes(R.color.colorPrimary,R.color.colorAccent);
                    mStates.put(pos, true);
                } else {           
                    holder.mFavourite.setLiked(false);
                    mStates.put(pos, false);
                }

That code isnt running everytime the button is clicked?

**Comment by salihyalcin** [comment]
mStates is SparseBooleanArray that I'm storing list views items state. Here is the whole code block .. 

```
 holder.mFavourite.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                int pos = notFilteredItems.indexOf(filteredlist);


                if (!mStates.get(pos)) {
                    holder.mFavourite.setLiked(true);
                    holder.mFavourite.setCircleEndColorRes(R.color.colorAccent);
                    holder.mFavourite.setExplodingDotColorsRes(R.color.colorPrimary,R.color.colorAccent);
                    mStates.put(pos, true);
                    editor.putString(String.valueOf(pos), filteredlist.getBolumAdi());

                } else {
                 
                    holder.mFavourite.setLiked(false);
                    mStates.put(pos, false);
                    editor.remove(String.valueOf(pos));

                }
                editor.apply();
            }
        });
```

**Comment by jd-alexander** [comment]
Ok the problem is you are manually setting setLiked which is causing the animation not work I presume. So remove holder.mFavourite.setLiked(true) and holder.mFavourite.setLiked(false); The button itself handles that. Also it's best if you use the OnLikeListener instead on onClickListener like the example. https://github.com/jd-alexander/LikeButton/blob/master/app/src/main/java/com/like/example/MainActivity.java

Here's the documentation for that event listener https://github.com/jd-alexander/LikeButton#like-event-listener

**Comment by jd-alexander** [comment]
You should only use setLiked when you are intially inflating your adapter's view. You shouldnt use it in the onClick function. That will mess up everything. The library handles the liked and unliked state. All you have to do is listen using the setOnLikeListener .More on the button state https://github.com/jd-alexander/LikeButton#button-state

**Comment by jd-alexander** [comment]
When you are usiing setOnClickListener you are somewhat overriding code that's in the library that's using that click event as a trigger for the animation so the listen for events simply use the listener that's bundled.

**Comment by salihyalcin** [comment]
But I have to control my mStates SparseBoolen array ? Can I control my Array using your like listener ?

**Comment by jd-alexander** [comment]
Yes you can simply set the on like listener and then set the boolean value based on if the item has been liked or not.

```java
 holder.mFavourite.setOnLikeListener(new OnLikeListener(){
@Override
    public void liked() {
       mStates.put(pos, true);
    }
```

    @Override
    public void unLiked() {
        mStates.put(pos, false);
    }
});

**Comment by salihyalcin** [comment]
You're right but I have to control mStates array before liking and unliking stuff.

**Comment by jd-alexander** [comment]
You could do the similar code, just not in the listener but in the adapter view generation code. So something like this. 

                if (mStates.get(pos)) {
                    holder.mFavourite.setLiked(true);

                } else {

                    holder.mFavourite.setLiked(false);
                  
                }

**Comment by jd-alexander** [comment]
Based on my understanding you should be storing the state of these items either via a web service or locally in a database or shared preferences. When the adapter is creating the view you should then get if the item was liked or not using a boolean value you stored which you then set the inital button state to liked or unliked. When you perform a click you now use the on like listener to modify the state of data layer that's storing the objects related to your view.
